### PR TITLE
bump up to webpack 5

### PR DESCRIPTION
**Pull request by blowery**
Updates the plugin to work for webpack@5. 
- use RawSource instead of jamming the content into a chunk
- use the new chunks api (now a Set instead of an array)
- use the new assets processing hooks
- redo tests to use webpack instead of mocking it out

### CODE REVIEW

File: .prettierrc [hunk 1]
```diff
@@ -0,0 +1,5 @@
+{
```
Comment (blowery) [inline .prettierrc:1]: match internal Tumblr prettier config

File: package.json [hunk 2]
```diff
@@ -25,11 +25,13 @@
   "license": "MIT",
   "dependencies": {
     "md5": "^2.2.1",
-    "webpack": "^4.39.1"
+    "webpack": "^5.0.0"
```
Comment (sirreal) [inline package.json:30]: This should probably be a peer (for users) and dev (for testing) dependency, not a regular dependency.

**Review by sirreal — COMMENTED** [review]
One note, I'll give a full review soon.

File: src/__tests__/plugin.tests.js [hunk 3]
```diff
@@ -0,0 +1,242 @@
...
+    return allJsonAssets[0];
+  };
+
+  const getManifestFromCompilationAssets = (compilationAssets) => {
+    const manifestFileName = getManifestFileNameFromCompilationAssets(compilationAssets);
+    // eslint-disable-next-line security/detect-object-injection
+    return compilationAssets[manifestFileName];
+  };
+
+  const distPath = path.join(__dirname, '..', '..', 'dist');
```
Comment (sirreal) [inline src/__tests__/plugin.tests.js]: Minor: It's surprising to have a git ignored top-level dist directory that doesn't contain compiled files for distribution. If it's just changing this line and .gitignore, can we use a more clear directory name, maybe even hidden, like `.test-output`?

File: src/plugin.js [hunk 4]
```diff
@@ -158,68 +158,73 @@ class WebAppManifestPlugin {
...
+            const hash = md5(content).substring(0, 8);
+            filename = `${normalizedDestination}/manifest-${hash}.json`;
+          } else {
+            filename = `${normalizedDestination}/manifest.json`;
+          }
+
+          /*
+          This adds the app manifest as an asset to Webpack.
+          */
+          // eslint-disable-next-line security/detect-object-injection, no-param-reassign
```
Comment (sirreal) [inline src/plugin.js]: Obsolete eslint disable
```suggestion
```

File: package.json (hunk 2 above)
Comment (sirreal) [reply package.json:30]: Pushed this in f88d2ff94ad1adfa4c20649d767c628f5fff117a

File: package.json [hunk 5]
```diff
@@ -12,7 +12,7 @@
     "url": "https://github.com/tumblr/webpack-web-app-manifest-plugin.git"
   },
   "engines": {
-    "node": ">=8.0.0"
+    "node": ">=10.13.0"
```
Comment (sirreal) [inline package.json:15]: Match: https://github.com/webpack/webpack/blob/610f36817af1f1c6929786d68f734ac582af1757/package.json#L107

File: src/__tests__/plugin.tests.js [hunk 6]
```diff
@@ -0,0 +1,242 @@
+const WebAppManifestPlugin = require('webpack-web-app-manifest-plugin');
+const { Chunk, webpack } = require('webpack');
+const path = require('path');
+const fs = require('fs/promises');
+
+describe('webpack-web-app-manifest-plugin', () => {
+  const makeCompiler = (assets = [], publicPath = '') => ({
+    assets: assets.reduce((memo, assetName) => {
+      // eslint-disable-next-line security/detect-object-injection, no-param-reassign
```
Comment (sirreal) [inline src/__tests__/plugin.tests.js]: This project doesn't seem to have eslint at all. Should we drop these?

File: src/__tests__/plugin.tests.js [hunk 7]
```diff
@@ -0,0 +1,242 @@
+const WebAppManifestPlugin = require('webpack-web-app-manifest-plugin');
+const { Chunk, webpack } = require('webpack');
```
Comment (sirreal) [inline src/__tests__/plugin.tests.js]: Unused
```suggestion
const { webpack } = require('webpack');
```

**Comment by sirreal** [comment]
I added an `.npmignore` to ensure we don't ship testing .png files in the packaged plugin:

Before

```
npm publish --dry-run
npm notice
npm notice 📦  webpack-web-app-manifest-plugin@3.0.2
npm notice === Tarball Contents ===
npm notice 664B   CONTRIBUTING.md
npm notice 1.1kB  LICENSE
npm notice 3.0kB  README.md
npm notice 73B    .prettierrc
npm notice 57.2kB assets/logo.gif
npm notice 45B    index.js
npm notice 1.5kB  package.json
npm notice 3.9kB  src/__tests__/assets/manifest/definitely_not_a_manifest_icon.png
npm notice 3.9kB  src/__tests__/assets/manifest/icon_192.png
npm notice 11.8kB src/__tests__/assets/manifest/icon_512.png
npm notice 3.9kB  src/__tests__/assets/manifest/this_is_a_manifest_icon.png
npm notice 6.6kB  src/__tests__/plugin.tests.js
npm notice 7.8kB  src/plugin.js
npm notice === Tarball Details ===
npm notice name:          webpack-web-app-manifest-plugin
npm notice version:       3.0.2
npm notice filename:      webpack-web-app-manifest-plugin-3.0.2.tgz
npm notice package size:  51.5 kB
npm notice unpacked size: 101.3 kB
npm notice shasum:        744bca12a5d6dac67dabeac8ca57c97a5980681b
npm notice integrity:     sha512-QIzgYjImGmbIE[...]yoLktHWk336Tw==
npm notice total files:   13
```

After

```
npm publish --dry-run
npm notice
npm notice 📦  webpack-web-app-manifest-plugin@3.0.2
npm notice === Tarball Contents ===
npm notice 664B   CONTRIBUTING.md
npm notice 1.1kB  LICENSE
npm notice 3.0kB  README.md
npm notice 57.2kB assets/logo.gif
npm notice 45B    index.js
npm notice 1.5kB  package.json
npm notice 7.8kB  src/plugin.js
npm notice === Tarball Details ===
npm notice name:          webpack-web-app-manifest-plugin
npm notice version:       3.0.2
npm notice filename:      webpack-web-app-manifest-plugin-3.0.2.tgz
npm notice package size:  34.6 kB
npm notice unpacked size: 71.2 kB
npm notice shasum:        c8d46b4f70a190a38b97e6c6bda41863cdf188bc
npm notice integrity:     sha512-8xeQaLfZ4WFCS[...]X95WvtkqPArCw==
npm notice total files:   7
npm notice
```

**Comment by sirreal** [comment]
Coverage had slipped, I've made a few changes to tests to restore 100% coverage as well as clean up after tests.
Do we need `selfHash`, or was that a testing-specific option? It seems like we could remove it, `md5` seems safe and simple to mock for tests.

Aside: Webpack has a build in `createHash` util we can use. Or we can just use the node `crypto` module which webpack uses under the hood. We could drop the `md5` dependency then.

**Review by sirreal — APPROVED** [review]
I've pushed  some minor changes that made sense to me, but overall this seems good. Webpack publishes types no, so I created a branch from this and typed it. There are a few issues but they're all preexisting. We could land that and publish types after this: #25

I think we can get this published under a `next` dist-tag and start to exercise it 👍

File: src/plugin.js (hunk 4 above)
Comment (sirreal) [reply src/plugin.js]: Do we want to use eslint? Maybe we can publish a tumblr config 🤔

**Comment by blowery** [comment]
@sirreal selfHash was really just for testing. Mocking out md5 would be fine; I think moving to letting webpack handle the hashing would be even better, but maybe that's a next major thing.

**Comment by sirreal** [comment]
I'm planning to take this over and land it.

**Pull request merged**